source/shared/cpp/ObjectModel/MarkDownParser.cpp

```diff
@@ -0,0 +1,430 @@
+bool MarkDownParser::IsLeftEmphasisDelimiter (void) 
+{
+    bool isLeftEmphasisDetecting = ((m_EmphasisState & ~EmphasisOn) && 
+                          m_delimiterCnts && 
+                          !isspace (*m_curPos) && 
+                          !(m_lookBehind == Alphanumeric && ispunct(*m_curPos)) && 
+                          !(m_lookBehind == Alphanumeric && m_currentDelimiterType == Underscore));
+    if (isLeftEmphasisDetecting)
+    {
+        m_leftLookUpTable.push_back(
+            Emphasis(m_delimiterCnts, m_currentWordIndex)); 
+        m_LeftEmphasisDetecting += m_delimiterCnts;
+    }
+    return isLeftEmphasisDetecting;
+}
+
+bool MarkDownParser::IsRightEmphasisDelimiter (void)
+{
+    bool isRightEmphasisDetected = false;
+    if (isspace (*m_curPos) && 
+       (m_lookBehind != WhiteSpace) && 
+       (m_checkLookAhead || m_checkIntraWord || m_currentDelimiterType == Asterisk))
+    {        
+        isRightEmphasisDetected  = true;
+    }
+
+    if (isalnum(*m_curPos) && m_lookBehind != WhiteSpace)
+    {
+        if (!m_checkLookAhead && !m_checkIntraWord)
+        {
+            isRightEmphasisDetected  = true;
+        }
+
+        if (m_checkLookAhead || m_checkIntraWord)
+            return m_isRightEmphasisDetecting = false;
+    }
+
+    if (ispunct(*m_curPos) && m_lookBehind != WhiteSpace)
+    {
+        if ((m_EmphasisState & EmphasisOff) || 
+           ((m_EmphasisState & EmphasisOn) && (m_curPos + 1 == m_textEnd))) 
+        {
+            isRightEmphasisDetected  = true;
+        }
+    }
+
+    if (isRightEmphasisDetected)
+    {
+        if (m_LeftEmphasisDetecting)
+        {
+            m_rightLookUpTable.push_back(
+                Emphasis(m_delimiterCnts, m_currentWordIndex)); 
+            m_LeftEmphasisDetecting -= m_delimiterCnts;
+            return m_isRightEmphasisDetecting = true;
+        }
+    }
+
+    return m_isRightEmphasisDetecting = false;
+}
+
+void MarkDownParser::SetText(const std::string & txt)
+{
+    m_text = txt;
+}
+
+bool MarkDownParser::IsMarkDownDelimiter (char ch)
+{
+    return ((ch == '*' || ch == '_') && (m_lookBehind != Escape));
+}
+
+// Updates Emphasis MarkDown State
+void MarkDownParser::UpdateState (void)
+{
+    if (IsMarkDownDelimiter (*m_curPos))
+    {
+        m_EmphasisState = (m_EmphasisState & (EmphasisOff))?
+            EmphasisDetected : EmphasisDetecting;
+    }
+    else
+    {
+        m_EmphasisState = (m_EmphasisState & EmphasisOn)?
+            EmphasisEnded : EmphasisNotDetected;
+    }
+
+}
+
+void MarkDownParser::GenSymbolTable (void)
+{
+    while (m_curPos < m_textEnd)
+    {
+        UpdateState ();
+
+        if (m_EmphasisState & EmphasisOn)
+        {
+            DelimiterType delimiterType = (*m_curPos == '*')? Asterisk : Underscore;
+
+            if (m_curPos != m_textBegin && (m_EmphasisState & EmphasisDetected))
+            {
+                m_tokenizedString.push_back("");
+                ++m_currentWordIndex;
+            }
```

Comment:
I realize that there are a lot of edge cases and complexities to this markdown parsing, but I'm finding this code generally difficult to read.  It would help both me as a code reviewer and any future developers who need to read this code if we could spend some time to make this code a little easier to understand.

More comments would help a lot:
- Since this is fairly non-intuitive code, a description at the top describing the overall functioning of the state machine at a high level might be helpful to give some context.
- Function level comments.  What's a symbol table and what is entailed in generating it?  
- Comments on blocks of code.  I can see that when we're in the EmphasisDetected state, but not at the beginning of the text, we're adding an empty string to the tokenized string and incrementing the word index.  I don't immediately know why.  What does adding an empty string to the tokenized string represent?  Why do we not need to do this if we're at the beginning of the string?  Why does this state indicate a new "Word"?

In addition to comments, think about how you're naming variables and functions.  In general, more descriptive is almost always better.  Also think about the meaning of the names you're giving them. 
This function is doing a lot of things, but it's not obviously updating or creating any variables with names like "symbol table".  A function comment will help, but making the code itself more obvious is even better.  Should this actually be GenerateTokenizedString?  Should some of the things being updated be more obviously named as symbol tables?  Is m_currentWordIndex actually about "words"?

Lastly, if you can thing of any ways that the structure of the code itself can be simplified or clarified that is the most helpful thing of all.  It may be the nature of the beast that this is complex code, but it's something to think about.  For example.  
-There are a lot of member variables representing state.  Do they all need to be member variables, or could they be local to the part of the code that needs them?  Ultimately there's only one function on the outside of this thing, so theoretically we shouldn't need to maintain a lot of state.  Declaring the variables close to where they're used makes the code easier to understand since we can get context on how they're used near the place where they're defined.  (P.S., what happens if you call TransformToHtml more than once?  does the state get reset?)
- Breaking out things like this into separate functions from TransformToHtml like you've done here helps, so this is great.  Are there any more parts of TransformToHtml that would make sense as helper functions?

I love that you've written so many tests for this thing so we have confidence that it works well, but code that's tricky makes me nervous even when we know it works, so the clearer we can be in the code itself the better.

Reply:
Thanks for the great advice!

---
In reply to: [154425566](https://github.com/Microsoft/AdaptiveCards/pull/1003#discussion_r154425566) [](ancestors = 154425566)

Reply:
i'm working on init method, and will remove unnecessary member variables and use locals instead to increase readability. 

meanwhile, i will refactor to make it more readable. rules themselves were very complex, so it was difficult to come up with abstraction that simplifies the rules, but i do understand the rules better, so I will attempt to make it clearer . 

---
In reply to: [154427017](https://github.com/Microsoft/AdaptiveCards/pull/1003#discussion_r154427017) [](ancestors = 154427017,154425566)